src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Instance/InstanceDiscovery.cs

```diff
@@ -0,0 +1,154 @@
+            new ConcurrentDictionary<string, InstanceDiscoveryMetadataEntry>();
+
+        private static SemaphoreSlim semaphore = new SemaphoreSlim(1, 1);
+
+        public static async Task<InstanceDiscoveryMetadataEntry> GetMetadataEntry(string host, bool validateAuthority,
+            CallState callState)
+        {
+            InstanceDiscoveryMetadataEntry entry = null;
+            if (!InstanceCache.TryGetValue(host, out entry))
+            {
+                await semaphore.WaitAsync().ConfigureAwait(false); // SemaphoreSlim.WaitAsync() will not block current thread
+                try
+                {
+                    if (!InstanceCache.TryGetValue(host, out entry))
+                    {
+                        await DiscoverAsync(host, validateAuthority, callState).ConfigureAwait(false);
+                        InstanceCache.TryGetValue(host, out entry);
+                    }
+                }
+                finally
+                {
+                    semaphore.Release();
+                }
+            }
+
+            return entry;
+        }
+
+        // No return value. Modifies InstanceCache directly.
+        private static async Task DiscoverAsync(string host, bool validateAuthority, CallState callState)
+        {
+            string tentativeAuthorizeEndpoint = AuthorizeEndpointTemplate.Replace("{host}", host);
+            string instanceDiscoveryEndpoint =
+                ("https://{host}/common/discovery/instance?api-version=1.1&authorization_endpoint=" +
+                 tentativeAuthorizeEndpoint)
+                .Replace("{host}", WhitelistedAuthorities.Contains(host) ? host : DefaultTrustedAuthority);
+            var client = new AdalHttpClient(instanceDiscoveryEndpoint, callState);
+            InstanceDiscoveryResponse discoveryResponse = null;
+            try
+            {
+                discoveryResponse = await client.GetResponseAsync<InstanceDiscoveryResponse>().ConfigureAwait(false);
+                if (validateAuthority && discoveryResponse.TenantDiscoveryEndpoint == null)
+                {
+                    // hard stop here
+                    throw new AdalException(AdalError.AuthorityNotInValidList);
+                }
+            }
+            catch (AdalServiceException ex)
+            {
+                if (validateAuthority)
+                {
+                    // hard stop here
+                    throw new AdalException(
+                        (ex.ErrorCode == "invalid_instance")
+                            ? AdalError.AuthorityNotInValidList
+                            : AdalError.AuthorityValidationFailed, ex);
+                }
+            }
+
+            foreach (var entry in discoveryResponse?.Metadata ?? Enumerable.Empty<InstanceDiscoveryMetadataEntry>())
+            {
+                foreach (var alias in entry?.Aliases ?? Enumerable.Empty<string>())
```

Comment:
minor nit: alias is a C# keyword. Granted it's out of context here and compiler shouldn't be confused, but generally we avoid naming variables to be exactly the same as keywords.

Reply:
Good to know `alias` is one of the ["contextual keyword"](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/index). Not even ReSharper gives a ~sh@t~ warning on that. I'll change it to _alias in next commit.